powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/VaultUnlockServiceBehavior.java

```diff
@@ -70,15 +104,104 @@ public void setServerPrivateKeyConverter(ServerPrivateKeyConverter serverPrivate
-     * @param isSignatureValid       Information about validity of the signature.
-     * @param keyConversionUtilities Key conversion utilities.
+     * @param applicationKey         Application key.
+     * @param signature              PowerAuth signature.
+     * @param signatureType          PowerAuth signature type.
+     * @param cryptogram             ECIES cryptogram.
+     * @param keyConversion          Key conversion utilities.
      * @return Vault unlock response with a properly encrypted vault unlock key.
      * @throws InvalidKeySpecException In case invalid key is provided.
      * @throws InvalidKeyException     In case invalid key is provided.
      * @throws GenericServiceException In case server private key decryption fails.
      */
-    public VaultUnlockResponse unlockVault(String activationId, boolean isSignatureValid, CryptoProviderUtil keyConversionUtilities) throws InvalidKeySpecException, InvalidKeyException, GenericServiceException {
-        throw new IllegalStateException("Not implemented yet");
+    public VaultUnlockResponse unlockVault(String activationId, String applicationKey, String signature, SignatureType signatureType,
+                                           String signedData, EciesCryptogram cryptogram, CryptoProviderUtil keyConversion)
+            throws GenericServiceException, InvalidKeySpecException, InvalidKeyException, EciesException, UnsupportedEncodingException, JsonProcessingException {
+
+        // Lookup the activation
+        final ActivationRecordEntity activation = repositoryCatalogue.getActivationRepository().findActivation(activationId);
+
+        // Check if the activation is in correct state
+        if (activation == null || !ActivationStatus.ACTIVE.equals(activation.getActivationStatus())) {
+            // Return response with invalid signature flag when activation is not valid
+            VaultUnlockResponse response = new VaultUnlockResponse();
+            response.setSignatureValid(false);
+            return response;
+        }
+
+        // Get the server private key, decrypt it if required
+        final String serverPrivateKeyFromEntity = activation.getServerPrivateKeyBase64();
+        final KeyEncryptionMode serverPrivateKeyEncryptionMode = activation.getServerPrivateKeyEncryption();
+        final String serverPrivateKeyBase64 = serverPrivateKeyConverter.fromDBValue(serverPrivateKeyEncryptionMode, serverPrivateKeyFromEntity, activation.getUserId(), activation.getActivationId());
+        byte[] serverPrivateKeyBytes = BaseEncoding.base64().decode(serverPrivateKeyBase64);
+        final PrivateKey serverPrivateKey = keyConversion.convertBytesToPrivateKey(serverPrivateKeyBytes);
+
+        // Get application secret and transport key used in sharedInfo2 parameter of ECIES
+        final ApplicationVersionEntity applicationVersion = repositoryCatalogue.getApplicationVersionRepository().findByApplicationKey(applicationKey);
```

Comment:
Does this version exist (missing `null` check, NPE can be thrown)? Is it supported (this would probably fail a couple of lines above when validating the signature, but I still prefer doing the check here)?

Reply:
I will add the checks, thanks.